### PR TITLE
audio interleaver

### PR DESCRIPTION
**Pull request by agga1**
Membrane element for interleaving 2 mono streams into stereo

### CODE REVIEW

File: lib/membrane_audio_interleaver.ex [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+      end
+    end
+  end
+
+  @impl true
+  def handle_caps(pad, caps, _context, state) do
+    case state.caps do
+      nil ->
+        state = %{state | caps: caps}
+        {{:ok, caps: {:output, %{caps | channels: 2}}, redemand: :output}, state}
```
Comment (sheldak) [inline lib/membrane_audio_interleaver.ex:183]: why 2 instead of `state.channels`?

File: lib/membrane_audio_interleaver.ex [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_context, %{caps: nil} = state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _context, %{channels: channels} = state) do
+    do_handle_demand(div(size, channels), state)
```
Comment (sheldak) [inline lib/membrane_audio_interleaver.ex:100]: I am not sure if `div` is enough here. If `div(size, channels) < size / channels`, demand will not be fulfilled. Additionally, `div(size, channels)` is not aware of sample size, so the element may receive only a fragment of the last sample and it will be unable to perform interleaving on that last sample (so less to send).

File: lib/membrane_audio_interleaver.ex [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+          end
+        )
+
+      {{:ok, demands}, state}
+    end
+  end
+
+  # try to interleave channels and formulate proper element callback return message
+  defp interleave_and_return(state) do
+    case(try_interleave(state)) do
```
Comment (sheldak) [inline lib/membrane_audio_interleaver.ex:220]: I didn't know that `case` can be used in that way! But still `case try_interleave(state) do` is a standard way which I would stick to

File: lib/membrane_audio_interleaver.ex [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    end
+  end
+
+  # interleave channels only if all queues are long enough (have at least 'sample_size' size)
+  defp try_interleave(%{caps: caps, pads: pads, order: order} = state) do
+    sample_size = Caps.sample_size(caps)
+
+    min_length =
+      min_queue_length(pads)
+      |> trunc_to_whole_samples(sample_size)
```
Comment (sheldak) [inline lib/membrane_audio_interleaver.ex:233]: It is a good practice to avoid using pipe operator just once

```suggestion
    min_length =
      pads
      |> min_queue_length()
      |> trunc_to_whole_samples(sample_size)
```

File: lib/membrane_audio_interleaver.ex [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    rest = rem(size, sample_size)
+    size - rest
+  end
+
+  defp any_finished?(pads, sample_size) do
+    pads
+    |> Enum.any?(fn
+      {_pad, %{queue: queue, stream_ended: true}} when byte_size(queue) < sample_size -> true
+      _entry -> false
+    end)
```
Comment (sheldak) [inline lib/membrane_audio_interleaver.ex:269]: - Using pipe operator once should be avoided.
- Ended stream and `byte_size(queue) < sample_size` means that the input is invalid, so maybe it would be better to raise error in such situation

**Review by sheldak — CHANGES_REQUESTED** [review]

File: lib/membrane_audio_interleaver.ex [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ],
+              order: [
+                type: :list,
+                spec: list(integer),
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:40]: ok

File: lib/membrane_audio_interleaver.ex [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    caps: Caps
+
+  @impl true
+  def handle_init(%__MODULE__{} = options) do
+    state =
+      options
+      |> Map.from_struct()
+      |> Map.put(:pads, %{})
+      |> Map.put(:finished, false)
+      |> Map.put(:channels, length(options.order))
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:65]: done

File: lib/membrane_audio_interleaver.ex [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+                """,
+                default: 2048
+              ],
+              order: [
+                type: :list,
+                spec: list(integer),
+                description: """
+                Order in which channels should be interleaved
+                """,
+                default: nil
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:44]: done

File: lib/membrane_audio_interleaver.ex (hunk 1 above)
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:183]: you're right, fixed

File: lib/membrane_audio_interleaver.ex (hunk 4 above)
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:233]: fixed

File: lib/membrane_audio_interleaver.ex (hunk 3 above)
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:220]: you mean the parentheses?

Comment (agga1) [reply lib/membrane_audio_interleaver.ex]: ok

File: lib/membrane_audio_interleaver.ex (hunk 2 above)
Comment (agga1) [reply lib/membrane_audio_interleaver.ex]: fixed

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      [join_binaries(zipped_chunks) | acc]
+    end)
+    |> join_binaries()
+  end
+
+  @doc """
+  Split bitstring into chunks of `chunk_size`. Chunks are returned in reversed order.
+  """
+  @spec to_chunks_reversed(bitstring, pos_integer(), list()) :: list
+  def to_chunks_reversed(mbinary, chunk_size, acc \\ [])
```
Comment (sheldak) [inline lib/membrane_audio_mixer/do_interleave.ex:52]: what does `mbinary` mean?

File: lib/membrane_audio_interleaver.ex (hunk 2 above)
Comment (sheldak) [reply lib/membrane_audio_interleaver.ex:101]: So after discussion with Bartek it can stay as it is

File: lib/membrane_audio_mixer/do_interleave.ex (hunk 9 above)
Comment (agga1) [reply lib/membrane_audio_mixer/do_interleave.ex:52]: changed to binary

File: lib/membrane_audio_interleaver.ex (hunk 2 above)
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:101]: after discussion, I went back to div option

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    payload = interleave_binaries(payloads, sample_size)
+
+    {payload, Map.new(pads_list)}
+  end
+
+  @doc """
+  Interleave binaries, taking 'sample_size' bytes at a time.
+  """
+  @spec interleave_binaries([binary()], pos_integer()) :: any
+  def interleave_binaries(payloads, sample_size)
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_mixer/do_interleave.ex:35]: ok

File: lib/membrane_audio_interleaver.ex [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+  use Membrane.Filter
+  use Bunch
+
+  alias Membrane.AudioMixer.DoInterleave
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
+
+  require Membrane.Logger
+
+  def_options caps: [
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex]: ok

File: README.md [hunk 12]
```diff
@@ -18,25 +18,33 @@ Add the following line to your `deps` in `mix.exs`. Run `mix deps.get`.
...
 caps received from input pads have to be identical and match ones in element option (if that 
 option is different from `nil`).
 
+All inputs have to be added before starting the pipeline and should not be changed
+during mixer's work.
+
+Mixing and interleaving is tested only for integer audio formats.
+### Mixer
+
+Provided element adds samples from all pads and clips the result to the maximum value for given
```
Comment (sheldak) [inline README.md:33]: I think `provided` is less suitable when there are more than one element

```suggestion
The Mixer adds samples from all pads and clips the result to the maximum value for given 
```

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    {payloads, pads_list} = get_payloads(bytes_per_channel, pads_inorder)
+
+    payload = interleave_binaries(payloads, sample_size)
+
+    {payload, Map.new(pads_list)}
+  end
+
+  #  Interleave binaries, taking `sample_size` bytes at a time.
+  defp interleave_binaries(payloads, sample_size)
+
```
Comment (sheldak) [inline lib/membrane_audio_mixer/do_interleave.ex:33]: Why do we need empty function here?

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  #  Split bitstring into chunks of `chunk_size`. Chunks are returned in reversed order.
+  defp to_chunks_reversed(binary, chunk_size, acc \\ [])
+
+  defp to_chunks_reversed(binary, chunk_size, acc) when byte_size(binary) <= chunk_size do
+    [binary | acc]
+  end
+
+  defp to_chunks_reversed(binary, chunk_size, acc) do
+    <<chunk::binary-size(chunk_size), rest::bitstring>> = binary
```
Comment (sheldak) [inline lib/membrane_audio_mixer/do_interleave.ex:53]: When matching strings, form `x <> y` is preferred

```suggestion
    <<chunk::binary-size(chunk_size)>> <> rest = binary
```

File: test/interleaver/do_interleave_test.exs [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+          to_pad(2, <<90, 100, 110, 120, 130, 140, 150, 160>>)
+        ])
+
+      order = [1, 2]
+      bytes_per_channel = 4
+
+      interleaved1 = <<1, 90, 2, 100, 3, 110, 4, 120>>
+      interleaved2 = <<1, 2, 90, 100, 3, 4, 110, 120>>
+      interleaved3 = <<1, 2, 3, 90, 100, 110, 4, 120>>
+      interleaved4 = <<1, 2, 3, 4, 90, 100, 110, 120>>
```
Comment (sheldak) [inline test/interleaver/do_interleave_test.exs:30]: I think we can use list here and later use `Enum.map/2` and `Enum.each/2` to avoid repeats and numbers in the names.

**Review by sheldak — CHANGES_REQUESTED** [review]

File: lib/membrane_audio_mixer/do_interleave.ex (hunk 13 above)
Comment (agga1) [reply lib/membrane_audio_mixer/do_interleave.ex:33]: fixed

File: lib/membrane_audio_mixer/do_interleave.ex (hunk 14 above)
Comment (agga1) [reply lib/membrane_audio_mixer/do_interleave.ex:53]: ok

File: test/interleaver/do_interleave_test.exs (hunk 15 above)
Comment (agga1) [reply test/interleaver/do_interleave_test.exs:30]: ok, i refactored this test

File: test/interleaver/do_interleave_test.exs [hunk 16]
```diff
@@ -23,23 +23,20 @@ defmodule Membrane.DdInterleaveTest do
...
+      cases = [
+        {1, <<1, 90, 2, 100, 3, 110, 4, 120>>},
+        {2, <<1, 2, 90, 100, 3, 4, 110, 120>>},
+        {3, <<1, 2, 3, 90, 100, 110, 4, 120>>},
+        {4, <<1, 2, 3, 4, 90, 100, 110, 120>>}
+      ]
+
+      Enum.each(cases, fn {sample_size, expected_binary} ->
+        expected = {expected_binary, new_pads}
+        assert expected == DoInterleave.interleave(bytes_per_channel, sample_size, pads, order)
```
Comment (sheldak) [inline test/interleaver/do_interleave_test.exs:38]: expected values should be on the right side (except for pattern match)
```suggestion
        assert DoInterleave.interleave(bytes_per_channel, sample_size, pads, order) == expected
```

**Review by sheldak — APPROVED** [review]

File: test/interleaver/do_interleave_test.exs (hunk 16 above)
Comment (agga1) [reply test/interleaver/do_interleave_test.exs]: done

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+  defp to_chunks_reversed(binary, chunk_size, acc) do
+    <<chunk::binary-size(chunk_size)>> <> rest = binary
+    to_chunks_reversed(rest, chunk_size, [<<chunk::binary-size(chunk_size)>> | acc])
+  end
+
+  defp order_pads(pads, order) do
+    order
+    |> Enum.map(fn name -> {Membrane.Pad, :input, name} end)
+    |> Enum.map(fn pad -> {pad, pads[pad]} end)
```
Comment (bblaszkow06) [inline lib/membrane_audio_mixer/do_interleave.ex:58]: There's a helper for that
```suggestion
    |> Enum.map(fn name ->
      pad = Membrane.Pad.ref(:input, name)
      {pad, pads[pad]}
    end)
```
It's a macro, so it can be used as a pattern, but forces to `require Membrane.Pad`, e.g.
```
require Membrane.Pad
alias Membrane.Pad
Pad.ref(:input, name) = {Pad, :input, :left}
assert name == :left
```

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+  #  Split bitstring into chunks of `chunk_size`. Chunks are returned in reversed order.
+  defp to_chunks_reversed(binary, chunk_size, acc \\ [])
+
+  defp to_chunks_reversed(binary, chunk_size, acc) when byte_size(binary) <= chunk_size do
+    [binary | acc]
+  end
+
+  defp to_chunks_reversed(binary, chunk_size, acc) do
+    <<chunk::binary-size(chunk_size)>> <> rest = binary
```
Comment (bblaszkow06) [inline lib/membrane_audio_mixer/do_interleave.ex:51]: Since you don't need `binary`:
```suggestion
  defp to_chunks_reversed(<<chunk::binary-size(chunk_size), rest::binary>>, chunk_size, acc) do
```

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    to_chunks_reversed(rest, chunk_size, [<<chunk::binary-size(chunk_size)>> | acc])
```
Comment (bblaszkow06) [inline lib/membrane_audio_mixer/do_interleave.ex:52]: chunk is already a binary
```suggestion
    to_chunks_reversed(rest, chunk_size, [chunk | acc])
```

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+  defp interleave_binaries(payloads, sample_size) do
+    payloads
+    # split each channel's payload into `sample_size` chunks (channels order is reversed)
+    |> Enum.map(fn payload -> to_chunks_reversed(payload, sample_size) end)
+    # zip corresponding chunks of different channels and concatenate them (channels order is again reversed)
+    |> Enum.zip_reduce([], fn zipped_chunks, acc ->
+      [join_binaries(zipped_chunks) | acc]
+    end)
+    |> join_binaries()
+  end
```
Comment (bblaszkow06) [inline lib/membrane_audio_mixer/do_interleave.ex:41]: This seems overly complicated
1. Make sure `get_payloads` gets only a multiple of sample size.
2. Chunk the data into samples using [`Bunch.Binary.chunk_every`](https://hexdocs.pm/bunch/Bunch.Binary.html#chunk_every/2) (step 1. and 2. can be done using `Bunch.Binary.chunk_every_rem`)
3. Zip and join payloads: `Enum.zip_with(&Enum.join/1) |> Enum.join()` or `Enum.zip_with(& &1) |> Enum.flatten() |> Enum.join()` (not sure which one is better performance-wise, so choose more readable)
4. Recreate the pad map (like you do)

This probably gets rid of all the helpers in this file invalidating the comments there, but I'll leave them for educational purposes :wink:

File: test/interleaver/audio_interleaver_test.exs [hunk 21]
```diff
@@ -0,0 +1,163 @@
+defmodule Membrane.AudioInterleaverTest do
```
Comment (bblaszkow06) [inline test/interleaver/audio_interleaver_test.exs:1]: Haven't checked the tests yet, will check them in next review

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: lib/membrane_audio_interleaver.ex [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :on_request,
+    demand_unit: :bytes,
+    caps: Caps
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:55]: ok

File: lib/membrane_audio_interleaver.ex [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+
+  Each input pad should be identified with your custom id (using `via_in(Pad.ref(:input, your_example_id)` )
+  """
+
+  use Membrane.Filter
+  use Bunch
+
+  alias Membrane.AudioMixer.DoInterleave
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:16]: ok

File: lib/membrane_audio_interleaver.ex [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+  def handle_event(pad, event, _context, state) do
+    Membrane.Logger.debug("Received event #{inspect(event)} on pad #{inspect(pad)}")
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_process(
+        pad,
+        %Buffer{payload: payload} = _buffer,
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:165]: ok

File: lib/membrane_audio_interleaver.ex [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+              |> byte_size()
+              |> then(&max(0, size - &1))
+
+            {:demand, {pad, demand_size}}
+          end
+        )
+
+      {{:ok, demands}, state}
+    end
+  end
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:222]: ok

File: lib/membrane_audio_interleaver.ex [hunk 26]
```diff
@@ -0,0 +1,289 @@
...
+
+  @impl true
+  def handle_process(
+        pad,
+        %Buffer{payload: payload} = _buffer,
+        _context,
+        %{input_caps: input_caps} = state
+      ) do
+    if state.finished do
+      {:ok, state}
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:170]: ok

File: lib/membrane_audio_interleaver.ex [hunk 27]
```diff
@@ -0,0 +1,289 @@
...
+
+        do_handle_demand(size, state)
+    end
+  end
+
+  @impl true
+  def handle_end_of_stream(pad, _context, %{input_caps: input_caps} = state) do
+    if state.finished do
+      # end of stream already sent
+      {:ok, state}
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:130]: ok

File: lib/membrane_audio_interleaver.ex [hunk 28]
```diff
@@ -0,0 +1,289 @@
...
+
+  @impl true
+  def handle_demand(
+        :output,
+        buffers_count,
+        :buffers,
+        _context,
+        %{frames_per_buffer: frames, input_caps: input_caps} = state
+      ) do
+    case input_caps do
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:115]: ok

File: lib/membrane_audio_interleaver.ex [hunk 29]
```diff
@@ -0,0 +1,289 @@
...
+  def handle_end_of_stream(pad, _context, %{input_caps: input_caps} = state) do
+    if state.finished do
+      # end of stream already sent
+      {:ok, state}
+    else
+      sample_size = Caps.sample_size(input_caps)
+
+      state =
+        case Bunch.Access.get_in(state, [:pads, pad]) do
+          %{queue: queue} when byte_size(queue) < sample_size ->
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:136]: so I shouldn't consider the possibility that I received the number of bytes that isn't the multiple of sample_size?

File: lib/membrane_audio_interleaver.ex [hunk 30]
```diff
@@ -0,0 +1,289 @@
...
+  end
+
+  @impl true
+  def handle_pad_added(pad, _context, state) do
+    state =
+      Bunch.Access.put_in(
+        state,
+        [:pads, pad],
+        %{queue: <<>>, stream_ended: false}
+      )
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:78]: ok

File: lib/membrane_audio_mixer/do_interleave.ex [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+  @doc """
+  Order queues according to `order`, take `bytes_per_channel` from each queue
+  (all queues must be at least `bytes_per_channel` long),
+  and interleave them.
+  """
+  @spec interleave(pos_integer(), pos_integer(), %{}, [integer()]) :: {any(), map()}
+  def interleave(bytes_per_channel, _sample_size, pads, _order) when map_size(pads) == 1 do
+    [{pad, data}] = Map.to_list(pads)
+
+    <<payload::binary-size(bytes_per_channel)>> <> remaining_queue = data.queue
```
[the remark this answers is not in the document]
Comment (sheldak) [reply lib/membrane_audio_mixer/do_interleave.ex:16]: I have reviewed similar thing before, but in favor of the `<<payload::binary-..>> <> remaining_queue` syntax. For styling I often relaate to this guide: [elixir-style-guide](https://github.com/lexmag/elixir-style-guide#patterns-matching-binaries) and it suggests using `<>`

File: lib/membrane_audio_interleaver.ex (hunk 29 above)
[the remark this answers is not in the document]
Comment (bblaszkow06) [reply lib/membrane_audio_interleaver.ex:136]: Bear in mind this is at the end of the stream - if you received only part of the sample then something is broken - it should not happen. I guess if the queue is not empty you should warn about it - this might indicate the caps were wrong so you produced garbage or the input is corrupt and there were no parser to catch that

File: lib/membrane_audio_mixer/do_interleave.ex (hunk 31 above)
[the remark this answers is not in the document]
Comment (bblaszkow06) [reply lib/membrane_audio_mixer/do_interleave.ex:16]: Never seen that one and it's not something we followed. We mostly based on credo (https://github.com/rrrene/elixir-style-guide) which in turn was based on https://github.com/christopheradams/elixir_style_guide#strings. In the latter `<>` is advised for strings, but says nothing about binaries. Anyway, for this case both seem OK, but for longer matches (i.e. with ore patterns) I prefer the one I've suggested. However I'm open for discussion on that

File: lib/membrane_audio_interleaver.ex (hunk 29 above)
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_audio_interleaver.ex:136]: Pasting suggested code I get an error "(CompileError) undefined variable ^pad. No variable "pad" has been defined before the current pattern". I tried it both with and without ^.

File: lib/membrane_audio_interleaver.ex [hunk 32]
```diff
@@ -159,80 +159,71 @@ defmodule Membrane.AudioInterleaver do
...
   end
 
-  # try to interleave channels and formulate proper element callback return message
-  defp interleave_and_return(state) do
-    case try_interleave(state) do
-      :none -> {:ok, state}
-      {:finished, {buffer, state}} -> {{:ok, buffer: buffer, end_of_stream: :output}, state}
-      {:ok, {buffer, state}} -> {{:ok, buffer: buffer}, state}
-    end
+  # send demand to input pads where current queue is not long enough
```
Comment (bblaszkow06) [inline lib/membrane_audio_interleaver.ex:213]:
```suggestion
  # send demand to input pads that don't have a long enough queue
```

File: test/interleaver/do_interleave_test.exs [hunk 33]
```diff
@@ -29,7 +29,6 @@ defmodule Membrane.DdInterleaveTest do
       cases = [
         {1, <<1, 90, 2, 100, 3, 110, 4, 120>>},
         {2, <<1, 2, 90, 100, 3, 4, 110, 120>>},
-        {3, <<1, 2, 3, 90, 100, 110, 4, 120>>},
```
Comment (bblaszkow06) [inline test/interleaver/do_interleave_test.exs]: What happened here? What is the reason for this change?

Comment (agga1) [reply test/interleaver/do_interleave_test.exs]: in this test I take 4 bytes_per_channel for each checked case, so I would have to create another test for the case with sample_size 3, which I thought was unnecessary (previously 'interleave' didn't check if bytes_per_channel were multiple of sample_size as it was ensured by AudioInterleave element, and 'interleave' just interleaved shorter chunks at the end).

Comment (agga1) [reply test/interleaver/do_interleave_test.exs]: So I will add another test for sample size 3 so that the test are complete.

File: lib/membrane_audio_interleaver.ex [hunk 34]
```diff
@@ -69,13 +69,7 @@ defmodule Membrane.AudioInterleaver do
   end
 
   @impl true
-  def handle_pad_added(_pad, %{playback_state: playback_state} = _context, _state)
-      when playback_state != :stopped do
-    raise("All pads should be connected before starting the element!")
-  end
```
Comment (bblaszkow06) [inline lib/membrane_audio_interleaver.ex]: I obviously I've missed the `!` when making a suggestion and you handled that well. However, this explicit error seems like a good addition. You can add it after the existing clause

**Review by bblaszkow06 — APPROVED** [review]
